src/evaluators/BinaryExpression.js

```diff
@@ -60,7 +60,7 @@ export function getPureBinaryOperationResultType(
   function reportErrorIfNotPure(purityTest: (Realm, Value) => boolean, typeIfPure: typeof Value): typeof Value {
     let leftPure = purityTest(realm, lval);
     let rightPure = purityTest(realm, rval);
-    if (leftPure && rightPure) return typeIfPure;
+    if ((leftPure && rightPure) || realm.isInPureScope()) return typeIfPure;
```

Comment:
I think this makes sense but maybe it needs to be:

```js
if (leftPure && rightPure) return typeIfPure;

if (realm.isInPureScope()) {
  if (!leftPure) Leak.leakValue(lval);
  if (!rightPure) Leak.leakValue(rval);
  return typeIfPure;
}
```

Similar to this: https://github.com/facebook/prepack/blob/4f22fbae831e08804104f1b08bcd8b5ad00c44ee/src/values/AbstractObjectValue.js#L346

?

Because they might leak the object to `toString()` and `valueOf()`.

I'm struggling to come up with a case where that would matter though.

Reply:
Yes, and also evaluateWithPossibleThrowCompletion since it might throw.